pkg/vpa/vpa.go

```diff
@@ -162,69 +154,128 @@ func (r Reconciler) namespaceIsManaged(namespace *corev1.Namespace) bool {
 	return r.OnByDefault
 }
 
-func (r Reconciler) reconcileDeploymentsAndVPAs(nsName string, vpaNames, deploymentNames []string, vpaUpdateMode v1beta2.UpdateMode) error {
-	// Create any VPAs that need to be
-	vpaNeeded := utils.Difference(deploymentNames, vpaNames)
-	klog.V(3).Infof("Diff deployments, vpas: %v", vpaNeeded)
+func (r Reconciler) reconcileDeploymentsAndVPAs(nsName string, vpas []v1beta2.VerticalPodAutoscaler, deployments []appsv1.Deployment, vpaUpdateMode v1beta2.UpdateMode) error {
+	// these keys will eventually contain the leftover vpas that do not have a matching deployment associated
+	vpaHasAssociatedDeployment := map[string]bool{}
+	for _, deployment := range deployments {
+		var dvpa *v1beta2.VerticalPodAutoscaler
+		// search for the matching vpa (will have the same name)
+		for _, vpa := range vpas {
+			if deployment.Name == vpa.Name {
+				// found the vpa associated with this deployment
+				dvpa = &vpa
+				vpaHasAssociatedDeployment[dvpa.Name] = true
+				break
+			}
+		}
 
-	if len(vpaNeeded) == 0 {
-		klog.Info("All VPAs are in sync.")
-	} else if len(vpaNeeded) > 0 {
-		for _, vpaName := range vpaNeeded {
-			err := r.createVPA(nsName, vpaName, vpaUpdateMode)
+		// for logging
+		vpaName := "none"
+		if dvpa != nil {
+			vpaName = dvpa.Name
+		}
+		klog.V(2).Infof("Reconciling Namespace/%s for Deployment/%s with VPA/%s in vpa-update-mode=%s", nsName, deployment.Name, vpaName, vpaUpdateMode)
+		err := r.reconcileDeploymentAndVPA(nsName, deployment, dvpa, vpaUpdateMode)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, vpa := range vpas {
+		if _, ok := vpaHasAssociatedDeployment[vpa.Name]; !ok {
+			// these vpas do not have a matching deployment, delete them
+			klog.V(2).Infof("Deleting dangling VPA/%s in Namespace/%s", vpa.Name, nsName)
+			err := r.deleteVPA(nsName, vpa.Name)
 			if err != nil {
 				return err
 			}
 		}
 	}
 
-	// Now that this is one, we can delete any VPAs that don't have matching deployments.
-	vpaDelete := utils.Difference(vpaNames, deploymentNames)
-	klog.V(5).Infof("Diff vpas, deployments: %v", vpaDelete)
+	return nil
+}
+
+func (r Reconciler) reconcileDeploymentAndVPA(nsName string, deployment appsv1.Deployment, vpa *v1beta2.VerticalPodAutoscaler, vpaUpdateMode v1beta2.UpdateMode) error {
+	// if the Deployment is opted out of vpa scaling then use vpa-update-mode="off"
+	if val, ok := deployment.GetAnnotations()[deploymentExcludedAnnotation]; ok {
+		if val == "true" {
+			klog.V(5).Infof("Deployment/%s has opted out of VPA scaling and will use vpa-update-mode=off", deployment.Name)
```

Comment:
This is the section that confused me about the intended functionality of the opt-out annotation. (see above comment)